### PR TITLE
refactor(charts): unified tooltip system with mobile-first map fixes

### PR DESCRIPTION
**Pull request by benhammondmusic**
## Summary

- Unified tooltip system across all chart types (rate bars, stacked bars, bubble chart, choropleth map, trends) using a shared `useChartTooltip<T>` hook and `HetChartHoverTooltip` portal component, replacing scattered d3-DOM tooltip utilities and ad-hoc local state
- Fixed map tooltip on mobile: React portal escapes ancestor flex container; JS-computed position with viewport clamping gives every tap a consistent `min(320px, 60vw)` width regardless of position on screen
- Map tooltip shows "click to explore" hint on desktop hover (detected via `(pointer: coarse)` media query, immune to synthetic mouse events generated by touch) and a tappable "Explore" button on touch
- Fixed scroll persistence across all charts: bar charts, bubble chart, and choropleth map dismiss tooltip on `scroll`/`wheel`; click/touch handlers use `svgRef.contains()` and `closest()` instead of fragile tag-name checks
- Fixed multimap modal: `HetChartHoverTooltip` accepts `inModal` prop and uses the `zIndexMultimapModalTooltip` token (1300) so tooltips appear above the MUI Dialog backdrop
- Added `aria-label` and `role="img"` to rate bar chart bars; keyboard focus ring uses brand green with rounded corners and no iOS tap highlight
- Nearest-neighbor bisector clamping in trends chart prevents out-of-bounds date lookup; `parsedDates` memoized to avoid per-mousemove allocations; empty-dates guard prevents index -1 edge case
- `window.matchMedia` calls guarded with optional chaining for safety in test/SSR environments
- Tooltip panel uses 97% opacity (`bg-alt-white/[0.97]`) for a subtle see-through effect; backdrop-blur removed (caused GPU compositing stutter when tooltip moved)

## Root Cause / Motivation

Charts previously used four different tooltip approaches (d3-DOM mutation in choropleth, local state + absolute positioning in bar charts, a custom local Tooltip component in bubble chart, and MUI Tooltip in trends). None handled viewport clamping, scroll dismissal, or touch/mouse disambiguation consistently. The map tooltip was the worst: `position: fixed` inside a flex ancestor caused iOS Safari to cap element width to the remaining right-side viewport space, producing narrow, clipped tooltips for states in the right half of the map.

## Test plan

- [x] TypeScript passes (`npx tsc --noEmit`)
- [x] Biome passes (`npm run cleanup`)
- [x] Rate bar chart: bars have `aria-label` attributes (Playwright)
- [x] Map tooltip: hover shows "click to explore" hint text (Playwright)
- [x] Trends chart: tooltip stable at chart edges — bisector clamp + nearest-neighbor (Playwright)
- [x] Tooltip stays within viewport bounds on map hover (Playwright)
- [x] Rate bar chart: hover tooltip appears and dismisses on mouseout (manual)
- [x] Rate bar chart: Tab through bars, verify each `aria-label` is announced by VoiceOver/NVDA
- [x] Keyboard nav: Tab to a bar, focus ring is brand green with rounded corners; no blue iOS tap highlight on touch
- [x] Stacked bar chart: tap a bar (mobile), tooltip appears; tap again, tooltip repositions; tap outside, tooltip dismisses
- [x] Map: tap a state on mobile — tooltip is consistent width regardless of tap position (left, right, center)
- [x] Map: tap a state on mobile — "Explore" button navigates to the state report
- [x] Map: hover a state on desktop — "click to explore" shows inline italic; clicking the region navigates without tooltip flicker
- [x] Multimap modal: hover a state in the compare-geographies view — tooltip appears above modal overlay
- [x] Bubble chart: screen reader browse mode walks all circles, reads location + both metric values + population
- [x] Card export: copy screenshot in dark mode — exported image has white background
- [x] Tooltip panel: 97% opacity visible against colored chart backgrounds; no motion stutter

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Desktop Screenshots (Hover Tooltips)

Screenshot 2026-06-08 at 2 51 54 PM
Screenshot 2026-06-08 at 2 51 59 PM
Screenshot 2026-06-08 at 2 52 16 PM
Screenshot 2026-06-08 at 2 52 32 PM
Screenshot 2026-06-08 at 2 52 39 PM
Screenshot 2026-06-08 at 2 52 56 PM
Screenshot 2026-06-08 at 2 53 22 PM
Screenshot 2026-06-08 at 2 53 26 PM
Screenshot 2026-06-08 at 2 53 31 PM
Screenshot 2026-06-08 at 2 53 33 PM
Screenshot 2026-06-08 at 2 53 37 PM
Screenshot 2026-06-08 at 2 53 43 PM
Screenshot 2026-06-08 at 2 53 48 PM
Screenshot 2026-06-08 at 2 53 52 PM
Screenshot 2026-06-08 at 2 54 22 PM
Screenshot 2026-06-08 at 2 54 26 PM

## Mobile Screenshots (Tap Tooltips)

image
image
image

### CODE REVIEW

File: frontend/src/charts/HetChartHoverTooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        left: `${x}px`,
+        top: `${y}px`,
+        transition: 'left 300ms ease-linear, top 300ms ease-linear',
+        transform: `translate(${flipLeft ? 'calc(-100% - 16px)' : '16px'}, ${flipUp ? 'calc(-100% - 16px)' : '16px'})`,
+      }}
+    >
+      {children}
+    </HetTooltipPanel>
+  )
+}
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/HetChartHoverTooltip.tsx:90]: Already addressed — the `animate` prop (default `false`) opts into the 300ms transition; only TrendsChart passes `animate`. Discrete charts (bar, bubble, map) snap instantly.

File: frontend/src/charts/HetChartHoverTooltip.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  return (
+    <div
+      role='tooltip'
+      className={`whitespace-nowrap rounded-sm border border-alt-gray border-solid bg-alt-white p-3 font-medium font-sans-text text-small${className ? ` ${className}` : ''}`}
+      style={style}
+    >
+      {children}
+    </div>
+  )
+}
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/HetChartHoverTooltip.tsx:23]: No `whitespace-nowrap` in the current code. Max-width is computed dynamically in `HetChartHoverTooltip` based on cursor position and viewport width (`maxWidth: flipLeft ? `${x - OFFSET - EDGE_PADDING}px` : `${vw - x - OFFSET - EDGE_PADDING}px``), so the tooltip always fits in the remaining viewport space.

File: frontend/src/charts/trendsChart/Index.tsx [hunk 3]
```diff
@@ -266,114 +209,83 @@ export function TrendsChart({
...
-        style={{
-          transform: `translate(${xScale(new Date(hoveredDate ?? ''))}px, ${
-            MARGIN.top
-          }px)`,
-          opacity: hoveredDate ? 1 : 0,
-        }}
+      <HetChartHoverTooltip
+        x={hoveredDate && tooltipPos ? tooltipPos.x : null}
+        y={hoveredDate && tooltipPos ? tooltipPos.y : null}
       >
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx:234]: Not applicable — we use the `animate` prop on `HetChartHoverTooltip` (already in place on TrendsChart) rather than a `transition` prop.

File: frontend/src/charts/CompareBubbleChart.tsx [hunk 4]
```diff
@@ -73,59 +74,48 @@ function weightedRegression(data: WeightedDataPoint[]): [number, number][] {
...
   const [resizeCardRef, width] = useResponsiveWidth()
   const svgRef = useRef<SVGSVGElement>(null)
   const [tooltipContent, setTooltipContent] =
-    useState<TooltipProps['content']>(null)
-  const [tooltipPosition, setTooltipPosition] = useState({ x: 0, y: 0 })
+    useState<BubbleTooltipContent | null>(null)
+  const [tooltipPosition, setTooltipPosition] = useState<{
+    x: number
+    y: number
+  } | null>(null)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/CompareBubbleChart.tsx]: Fixed — added `wheel` and `scroll` listeners to the existing window event useEffect, matching the choropleth map pattern (78fd978).

File: frontend/src/charts/BubbleChartTooltip.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      </div>
+      <div className='font-normal'>
+        {data.xLabel}: {data.xValue}
+      </div>
+      <div className='font-normal'>
+        {data.yLabel}: {data.yValue}
+      </div>
+      <div className='font-normal'>
+        Population: {data.population.toLocaleString()}
+      </div>
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/BubbleChartTooltip.tsx:25]: Fixed — added `?? 0` fallback before `.toLocaleString()`.

File: frontend/src/charts/choroplethMap/index.tsx [hunk 6]
```diff
@@ -79,18 +89,25 @@ const ChoroplethMap = ({
...
+  useEffect(() => {
+    window.addEventListener('wheel', hideTooltip)
+    window.addEventListener('click', hideTooltip)
+    window.addEventListener('touchmove', hideTooltip)
+    return () => {
+      window.removeEventListener('wheel', hideTooltip)
+      window.removeEventListener('click', hideTooltip)
+      window.removeEventListener('touchmove', hideTooltip)
     }
+  }, [hideTooltip])
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/choroplethMap/index.tsx]: Fixed — replaced `window.addEventListener('click', hideTooltip)` with a handler that skips path/circle elements and anything inside `[role="tooltip"]`, so mobile touchstart + synthetic click no longer cancels the tooltip.

File: frontend/src/charts/rateBarChart/Index.tsx [hunk 7]
```diff
@@ -95,10 +92,14 @@ export function RateBarChart(props: RateBarChartProps) {
   return (
     <div
       ref={containerRef}
-      onTouchStart={handleContainerTouch}
+      onTouchStart={(e) => {
+        if ((e.target as SVGElement).tagName !== 'path') hideTooltip()
+      }}
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/rateBarChart/Index.tsx:106]: Fixed — replaced `tagName !== 'path'` with `!target.closest('g[role="img"]')` so tapping text labels or any child of a bar group no longer dismisses the tooltip.

File: frontend/src/charts/stackedSharesBarChart/Index.tsx [hunk 8]
```diff
@@ -95,15 +97,21 @@ export function StackedBarChart(props: StackedBarChartProps) {
   return (
     <div
       ref={containerRef}
-      onTouchStart={handleContainerTouch}
+      onTouchStart={(e) => {
+        if ((e.target as SVGElement).tagName !== 'path') hideTooltip()
+      }}
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/stackedSharesBarChart/Index.tsx:111]: Fixed — same `closest('g[role="img"]')` approach used in the rate bar chart.

File: frontend/src/charts/CompareBubbleChart.tsx [hunk 9]
```diff
@@ -73,65 +77,40 @@ function weightedRegression(data: WeightedDataPoint[]): [number, number][] {
...
   const height = Math.min(
     isMd ? width * HEIGHT_WIDTH_RATIO : width / HEIGHT_WIDTH_RATIO,
     window.innerHeight * HEIGHT_WIDTH_RATIO,
   )
 
+  useEffect(() => {
+    const hideOnOutsideTouch = (e: TouchEvent) => {
+      if ((e.target as Element).tagName.toLowerCase() !== 'circle')
+        hideTooltip()
+    }
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/CompareBubbleChart.tsx:103]: Fixed — added optional chaining: `tagName?.toLowerCase()`.

File: frontend/src/charts/trendsChart/Index.tsx [hunk 10]
```diff
@@ -219,38 +170,33 @@ export function TrendsChart({
...
+        showTooltip(
+          nearestDate,
+          svgRect.left + xScale(new Date(nearestDate)),
+          svgRect.top + MARGIN.top,
+        )
+      }
     },
-    [dates, xScale],
+    [dates, xScale, MARGIN.top, showTooltip],
   )
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx:203]: Fixed — clamp idx to `dates.length - 1` and do a nearest-neighbor comparison between left and right boundary before selecting the date.

File: frontend/src/charts/choroplethMap/renderMap.tsx [hunk 11]
```diff
@@ -160,21 +145,17 @@ export const renderMap = (options: RenderMapOptions) => {
...
     })
-    .on('pointerdown', (event: any, d) => {
-      hideTooltips()
-      createEventHandler('pointerdown', mouseEventOptions)(event, d)
-    })
-    .on('mousemove', (event: any, d) => {
-      createEventHandler('mousemove', mouseEventOptions)(event, d)
+    .on('pointerdown', () => {
+      options.tooltipCallbacks.onHide()
     })
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/choroplethMap/renderMap.tsx]: Fixed — removed pointerdown hide handler; it caused the tooltip to vanish on desktop clicks since no mouseover fired afterward to restore it.

File: frontend/src/charts/choroplethMap/TerritoryCircles.tsx [hunk 12]
```diff
@@ -115,29 +110,21 @@ export default function TerritoryCircles(props: TerritoryCirclesProps) {
...
-          createTerritoryFeature(d.fips),
-        )(event, d)
-      })
-      .on('mousemove', (event: any, d) => {
-        createEventHandler('mousemove', mouseEventOptions, (d) =>
-          createTerritoryFeature(d.fips),
-        )(event, d)
+      .on('pointerdown', () => {
+        props.tooltipCallbacks.onHide()
       })
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/choroplethMap/TerritoryCircles.tsx]: Fixed — same pointerdown removal as renderMap.

File: frontend/src/charts/CompareBubbleChart.tsx [hunk 13]
```diff
@@ -73,65 +77,40 @@ function weightedRegression(data: WeightedDataPoint[]): [number, number][] {
...
   const height = Math.min(
     isMd ? width * HEIGHT_WIDTH_RATIO : width / HEIGHT_WIDTH_RATIO,
     window.innerHeight * HEIGHT_WIDTH_RATIO,
   )
 
+  useEffect(() => {
+    const hideOnOutsideTouch = (e: TouchEvent) => {
+      if ((e.target as Element).tagName?.toLowerCase() !== 'circle')
+        hideTooltip()
+    }
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/CompareBubbleChart.tsx:103]: Fixed — switched to svgRef.current.contains(e.target) so the check is scoped to this chart's SVG, not any circle on the page.

File: frontend/src/charts/choroplethMap/index.tsx [hunk 14]
```diff
@@ -79,18 +89,39 @@ const ChoroplethMap = ({
...
+    }
+    window.addEventListener('wheel', hideTooltip)
+    window.addEventListener('click', handleOutsideClick)
+    window.addEventListener('touchmove', hideTooltip)
+    return () => {
+      window.removeEventListener('wheel', hideTooltip)
+      window.removeEventListener('click', handleOutsideClick)
+      window.removeEventListener('touchmove', hideTooltip)
     }
+  }, [hideTooltip])
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/choroplethMap/index.tsx:124]: Fixed — using svgRef.current.contains for click handler and added passive scroll listener.

File: frontend/src/charts/trendsChart/Index.tsx [hunk 15]
```diff
@@ -125,81 +94,65 @@ export function TrendsChart({
...
   const dates = getDates(filteredData)
-  // Array of just amounts (y values)
   const amounts = getAmounts(filteredData)
 
-  /* Scales */
-
-  // define X and Y extents
   const xExtent: [Date, Date] | [undefined, undefined] = extent(
     dates.map((date) => new Date(date)),
   )
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx]: Fixed — parsedDates memoized with useMemo; reused in xExtent and handleMousemove.

File: frontend/src/charts/trendsChart/Index.tsx [hunk 16]
```diff
@@ -219,38 +170,47 @@ export function TrendsChart({
...
+        showTooltip(
+          nearestDate,
+          svgRect.left + xScale(new Date(nearestDate)),
+          svgRect.top + MARGIN.top,
+        )
+      }
     },
-    [dates, xScale],
+    [dates, xScale, MARGIN.top, showTooltip],
   )
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx:203]: Fixed — handleMousemove now uses the memoized parsedDates array; removed per-event Date allocations.

File: frontend/src/charts/HetChartHoverTooltip.tsx [hunk 17]
```diff
@@ -0,0 +1,84 @@
+import { createPortal } from 'react-dom'
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/HetChartHoverTooltip.tsx:1]: Not changing — reading window.innerWidth directly at render time is intentional. useViewportSize uses useState which returns stale values on initial renders, which was the root cause of the original flip/overflow bug.

File: frontend/src/charts/HetChartHoverTooltip.tsx [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+  children,
+  interactive = false,
+  animate = false,
+}: HetChartHoverTooltipProps) {
+  if (x === null || y === null) return null
+
+  // Read directly from window — avoids stale useState on initial renders,
+  // which caused flipLeft to be wrong and the tooltip to overflow off-screen.
+  const vw = window.innerWidth
+  const vh = window.innerHeight
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/HetChartHoverTooltip.tsx:51]: Not changing — same reason as the import comment above.

File: frontend/src/charts/CompareBubbleChart.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/CompareBubbleChart.tsx:103]: Already addressed — commit 94dd9ef: hideOnOutsideTouch uses svgRef.current.contains(e.target) so only taps outside the current SVG dismiss the tooltip.

File: frontend/src/charts/choroplethMap/index.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/choroplethMap/index.tsx:124]: Already addressed — the click handler uses svgRef.current.contains(target) and a scroll listener was added in the same commit.

File: frontend/src/charts/HetChartHoverTooltip.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/HetChartHoverTooltip.tsx:1]: Not applicable — useViewportSize was deleted (commit 0b55b40) as it was unused. window.innerWidth is read directly on each render, which is correct: the component re-renders on every mouse position update, so it always gets a fresh viewport measurement.

File: frontend/src/charts/HetChartHoverTooltip.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/HetChartHoverTooltip.tsx:51]: Not applicable — see reply above; the hook no longer exists.

File: frontend/src/charts/trendsChart/Index.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx]: Already done — parsedDates is memoized at line 118: useMemo(() => dates.map((d) => new Date(d)), [dates]).

File: frontend/src/charts/trendsChart/Index.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx:203]: Already done — the handler uses parsedDates from the memo (lines 175-196) and avoids any new Date() calls per event.

File: frontend/src/charts/trendsChart/Index.tsx [hunk 19]
```diff
@@ -219,38 +169,46 @@ export function TrendsChart({
...
-      const closestIdx = bisect.left(
-        dates.map((d) => new Date(d)),
-        invertedDate,
-      )
-      // set state to story hovered date
-      setHoveredDate(dates[closestIdx])
+      const svgRect = e.currentTarget.getBoundingClientRect()
+      const invertedDate = xScale.invert(e.clientX - svgRect.left)
+      const bisect = bisector((d: Date) => d)
+      let idx = bisect.left(parsedDates, invertedDate)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/trendsChart/Index.tsx:176]: Fixed — added `if (dates.length === 0) return` at the top of handleMousemove.

File: frontend/src/charts/choroplethMap/index.tsx [hunk 20]
```diff
@@ -228,6 +253,21 @@ const ChoroplethMap = ({
...
+        inModal={isMulti}
+      >
+        {mapTooltipData && (
+          <MapTooltipContent
+            data={mapTooltipData}
+            onExplore={updateFipsCallback}
+            isTouch={window.matchMedia('(pointer: coarse)').matches}
+          />
+        )}
+      </HetChartHoverTooltip>
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply frontend/src/charts/choroplethMap/index.tsx:272]: Fixed — replaced bare `window.matchMedia(...)` with `window.matchMedia?.(...) ?? false` optional chaining.

**Review by eriwarr — APPROVED** [review]
Awesome work on this refactor! The tool tips look so much better on mobile now!
